server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/ShardsBalancer.java

```diff
@@ -37,6 +37,8 @@ public abstract class ShardsBalancer {
      */
     abstract void balance();
 
+    abstract void tierShards();
```

Comment:
Can we keep these changes related to allocator in separate PR ? Before making these, we will need to introduce a mechanism to define a warm node. Search node doesn't seem to be the right construct as those are specific to read-only indices or for shards where only search requests are routed. Warm indices are mutable, so we don't want to move all its shards to the search nodes. We will need to probably differentiate data nodes based on the tier attributes like hot/warm and then use that to assign shards of a warm index. This in itself can be handled in a separate issue.

Reply:
removed this logic from this pr for now